op-supervisor/supervisor/backend/db/fromda/update.go

```diff
@@ -0,0 +1,111 @@
+package fromda
+
+import (
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/backend/db/entrydb"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+)
+
+func (db *DB) AddDerived(derivedFrom eth.BlockRef, derived eth.BlockRef) error {
+	db.rwLock.Lock()
+	defer db.rwLock.Unlock()
+
+	// If we don't have any entries yet, allow any block to start things off
+	if db.store.Size() == 0 {
+		link := LinkEntry{
+			derivedFrom: types.BlockSeal{
+				Hash:      derivedFrom.Hash,
+				Number:    derivedFrom.Number,
+				Timestamp: derivedFrom.Time,
+			},
+			derived: types.BlockSeal{
+				Hash:      derived.Hash,
+				Number:    derived.Number,
+				Timestamp: derived.Time,
+			},
+		}
+		e := link.encode()
+		if err := db.store.Append(e); err != nil {
+			return err
+		}
+		db.m.RecordDBDerivedEntryCount(db.store.Size())
+		return nil
+	}
+
+	lastDerivedFrom, lastDerived, err := db.latest()
+	if err != nil {
+		return err
+	}
+
+	if lastDerived.ID() == derived.ID() && lastDerivedFrom.ID() == derivedFrom.ID() {
```

Comment:
It's possible for these structures to still differ in Timestamp, right? Shall we error if those don't match?

```go
func (id L1BlockRef) ID() BlockID {
	return BlockID{
		Hash:   id.Hash,
		Number: id.Number,
	}
}
```